File: apps/Android/issueskit/src/main/kotlin/com/druware/issueskit/IssueDate.kt
package com.druware.issueskit

import java.time.Instant
import java.time.LocalDate
import java.time.OffsetDateTime
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import java.time.temporal.ChronoUnit

/**
 * Fixed date handling for the `YYYY-MM-DD` reported dates in the document.
 *
 * A single fixed-offset (UTC) calendar guarantees that parsing then formatting a date is a stable
 * identity regardless of the device's time zone, which is what keeps issue round trips lossless
 * and stops a "reported" day from shifting for users east or west of GMT.
 */
object IssueDate {

    private val dayFormatter: DateTimeFormatter =
        DateTimeFormatter.ofPattern("uuuu-MM-dd").withZone(ZoneOffset.UTC)

    /** Parses a `YYYY-MM-DD` string into a day-normalized instant, or `null` if malformed. */
    fun dateFrom(text: String): Instant? = try {
        LocalDate.parse(text.trimSpaces(), dayFormatter).atStartOfDay(ZoneOffset.UTC).toInstant()
    } catch (_: java.time.format.DateTimeParseException) {
        null
    }

    /** Formats an instant as `YYYY-MM-DD` in UTC. */
    fun stringFrom(date: Instant): String = dayFormatter.format(date)

    /** The current date normalized to the start of its day in UTC. */
    fun today(): Instant = Instant.now().truncatedTo(ChronoUnit.DAYS)
}

/**
 * ISO-8601 timestamps as written by the `.issues` format: UTC, whole seconds, `Z` suffix.
 *
 * Apple's `JSONEncoder.dateEncodingStrategy = .iso8601` truncates sub-second precision, so a
 * timestamp never survives a round trip finer than one second. This mirrors that exactly.
 */
internal object IssueTimestamp {

    private val formatter: DateTimeFormatter =
        DateTimeFormatter.ofPattern("uuuu-MM-dd'T'HH:mm:ss'Z'").withZone(ZoneOffset.UTC)

    fun format(instant: Instant): String = formatter.format(instant.truncatedTo(ChronoUnit.SECONDS))

    /** Accepts `Z` and numeric-offset forms, with or without fractional seconds. */
    fun parse(text: String): Instant? = try {
        Instant.parse(text)
    } catch (_: java.time.format.DateTimeParseException) {
        try {
            OffsetDateTime.parse(text).toInstant()
        } catch (_: java.time.format.DateTimeParseException) {
            null
        }
    }
}

/**
 * Trims the characters Swift's `CharacterSet.whitespaces` covers — horizontal tab plus Unicode
 * space separators — and deliberately not newlines, so the markdown parsers behave identically.
 */
internal fun String.trimSpaces(): String = trim { it.isIssueSpace() }

internal fun Char.isIssueSpace(): Boolean =
    this == '\t' || Character.getType(this) == Character.SPACE_SEPARATOR.toInt()

File: apps/Windows/src/IssuesKit/IssueDate.cs
using System.Globalization;

namespace IssuesKit;

/// <summary>
/// Fixed date handling for the <c>YYYY-MM-DD</c> reported dates in the document.
/// </summary>
/// <remarks>
/// A single invariant/UTC calendar and format guarantees that parsing then formatting a date is
/// a stable identity, which is what keeps issue round trips lossless.
/// </remarks>
public static class IssueDate
{
    private const string DayFormat = "yyyy-MM-dd";

    /// <summary>Parses a <c>YYYY-MM-DD</c> string into a day-normalized UTC date, or <c>null</c> if malformed.</summary>
    public static DateTimeOffset? Date(string? text)
    {
        if (text is null)
        {
            return null;
        }

        var parsed = DateTimeOffset.TryParseExact(
            text.Trim(),
            DayFormat,
            CultureInfo.InvariantCulture,
            DateTimeStyles.AssumeUniversal | DateTimeStyles.AdjustToUniversal,
            out var value);

        return parsed ? value : null;
    }

    /// <summary>Formats a date as <c>YYYY-MM-DD</c> in UTC.</summary>
    public static string String(DateTimeOffset date) =>
        date.ToUniversalTime().ToString(DayFormat, CultureInfo.InvariantCulture);

    /// <summary>The current date normalized to the start of its day in UTC.</summary>
    public static DateTimeOffset Today()
    {
        var now = DateTimeOffset.UtcNow;
        return new DateTimeOffset(now.Year, now.Month, now.Day, 0, 0, 0, TimeSpan.Zero);
    }
}
